File: odoo_surface_mcp/tools/intent.py
"""Layer 4 — Primary Intent tools: create, update, execute_action, archive,
post_message, schedule_activity."""
from typing import Any, Optional

from mcp.server.fastmcp import FastMCP

from odoo_surface_mcp.cache import Cache
from odoo_surface_mcp.odoo_client import OdooClient
from odoo_surface_mcp.tools.supporting import (
    _action_domain_context,
    _resolve_context,
    _safe_eval_dict,
    _view_field_names,
)


def register(mcp: FastMCP, client: OdooClient, cache: Cache) -> None:

    @mcp.tool(description=(
        "Create a new record. values: dict of field/value pairs (form-view fields only). "
        "Defaults are merged with provided values automatically. "
        "Pass action_id to include the action's context (e.g. default_partner_id). "
        "Pass context dict directly for wizard models whose context comes from an "
        "execute_action response (e.g. active_model, active_ids for payment wizards). "
        "Returns {id, display_name} or {error}."
    ))
    def create(
        model: str,
        values: dict,
        action_id: Optional[int] = None,
        context: Optional[dict] = None,
    ) -> dict:
        try:
            merged_ctx = _resolve_context(client, cache, action_id, context)
            new_id = client.execute(model, "create", values, context=merged_ctx)
            rows = client.execute(model, "read", [new_id], fields=["id", "display_name"])
            return rows[0] if rows else {"id": new_id}
        except Exception as exc:
            return {"error": str(exc)}

    @mcp.tool(description=(
        "Update fields on an existing record. values: {field: value, ...}. "
        "Writes both form-view fields and model fields not exposed in the form view "
        "(e.g. html content fields rendered by the website builder). "
        "Model-level readonly fields are silently skipped; the server enforces access rights. "
        "One2many / many2many fields accept Odoo Command tuples directly: "
        "  [[0,0,{vals}]]  — create and link a new line, "
        "  [[1,id,{vals}]] — update existing line by id, "
        "  [[2,id]]        — delete line by id, "
        "  [[6,0,[ids]]]   — replace the full set. "
        "Returns {success, updated_fields, non_form_fields} or {error}."
    ))
    def update(model: str, record_id: int, values: dict) -> dict:
        try:
            form_fields = _view_field_names(client, cache, model, "form")
            meta = client.execute(model, "fields_get", attributes=["readonly"])
            writable = {
                k: v for k, v in values.items()
                if k in meta and (
                    _is_odoo_command(v)
                    or not meta.get(k, {}).get("readonly", False)
                )
            }
            if not writable:
                return {"error": "No writable fields found in the provided values."}
            client.execute(model, "write", [record_id], writable)
            # Report form vs non-form writes so the caller knows which fields
            # are rendered outside the standard backend form view.
            form_written = [k for k in writable if k in form_fields]
            non_form_written = [k for k in writable if k not in form_fields]
            result: dict = {"success": True, "updated_fields": list(writable.keys())}
            if non_form_written:
                result["non_form_fields"] = non_form_written
            return result
        except Exception as exc:
            return {"error": str(exc)}

    @mcp.tool(description=(
        "Execute a button or server action on a record. "
        "action: button name (method) or label as shown in get_model_actions — "
        "e.g. 'action_confirm', 'Confirm', 'Privacy Lookup'. "
        "View buttons (type=object) call the method directly; server actions use ir.actions.server.run. "
        "Returns the Odoo action result, {success: true}, or {error}."
    ))
    def execute_action(model: str, record_id: int, action: str) -> dict:
        try:
            from odoo_surface_mcp.tools.discovery import _collect_model_actions
            action_map = _collect_model_actions(client, model)

            # 1. View buttons
            for btn in action_map.get("view_buttons", []):
                if btn.get("name") != action and btn.get("label") != action:
                    continue

                if btn.get("type") == "action":
                    # type="action" buttons carry a numeric ir.actions.* id as name.
                    # Load the action definition and inject active-record context —
                    # the agent (or a follow-up create+execute) handles the wizard.
                    action_id = int(btn["name"])
                    meta = client.execute(
                        "ir.actions.actions", "read", [action_id],
                        fields=["type", "name"],
                    )
                    if not meta:
                        return {"error": f"Action id {action_id} not found."}
                    action_type = meta[0].get("type", "ir.actions.act_window")
                    full = client.execute(action_type, "read", [action_id])
                    if not full:
                        return {"error": f"Could not load action {action_id}."}
                    action_def = dict(full[0])
                    ctx = _safe_eval_dict(action_def.get("context") or {})
                    ctx.update(_active_ctx(record_id, model))
                    action_def["context"] = ctx
                    return action_def

                # type="object" — use /web/dataset/call_button (browser equivalent).
                # The endpoint calls clean_action() on action-dict results and returns
                # False for everything else (True / None / non-action return values).
                result = client.http_call("/web/dataset/call_button", {
                    "model": model,
                    "method": btn["name"],
                    "args": [[record_id]],
                    "kwargs": {},
                })

                # False / None → button ran with no redirect; re-read the record.
                if result is False or result is None:
                    return _action_check(client, model, record_id)

                # Returned a clean action dict (e.g. opens a wizard) — pass through.
                return _normalise(result)

            # 2. Server actions — run via ir.actions.server with active context
            for sa in action_map.get("server_actions", []):
                if str(sa.get("id")) == action or sa.get("name") == action:
                    result = client.execute(
                        "ir.actions.server", "run", [sa["id"]],
                        context=_active_ctx(record_id, model),
                    )

                    if result is True or result is False or result is None:
                        return _action_check(client, model, record_id)

                    return _normalise(result)

            return {"error": (
                f"Action '{action}' not found on '{model}'. "
                "Use get_model_actions() to list available actions."
            )}
        except Exception as exc:
            return {"error": str(exc)}

    @mcp.tool(description=(
        "Archive (deactivate) a record by setting active=False. "
        "Only works on models that have an active field (most standard models do). "
        "Returns {success: true} or {error}."
    ))
    def archive(model: str, record_id: int) -> dict:
        try:
            meta = client.execute(model, "fields_get", attributes=["type"])
            if "active" not in meta:
                return {"error": f"Model '{model}' has no active field and cannot be archived."}
            client.execute(model, "write", [record_id], {"active": False})
            return {"success": True, "record_id": record_id, "active": False}
        except Exception as exc:
            return {"error": str(exc)}

    @mcp.tool(description=(
        "Post a message or internal note on a record (requires mail.thread). "
        "message_type: 'comment' (sent to followers, visible in chatter) "
        "or 'note' (internal log note, not emailed). "
        "Returns {message_id} or {error}."
    ))
    def post_message(
        model: str,
        record_id: int,
        body: str,
        message_type: str = "comment",
    ) -> dict:
        try:
            if message_type not in ("comment", "note"):
                return {"error": "message_type must be 'comment' (sent to followers) or 'note' (internal only)."}
            # In Odoo 17, message_post always uses message_type='comment'.
            # The comment vs internal-note distinction is made via subtype_xmlid:
            #   mail.mt_comment → visible message sent to followers
            #   mail.mt_note    → internal log note, not emailed
            msg_id = client.execute(
                model, "message_post", [record_id],
                body=body,
                message_type="comment",
                subtype_xmlid="mail.mt_note" if message_type == "note" else "mail.mt_comment",
            )
            return {"message_id": msg_id}
        except Exception as exc:
            return {"error": str(exc)}

    @mcp.tool(description=(
        "Schedule an activity on a record. "
        "activity_type: name of the activity type (e.g. 'To-Do', 'Email', 'Phone Call'). "
        "deadline: ISO date string YYYY-MM-DD. "
        "summary: short title. note: longer description (optional). "
        "assigned_user_id: who to assign (default: current user). "
        "Returns {activity_id, activity_type, deadline} or {error}."
    ))
    def schedule_activity(
        model: str,
        record_id: int,
        activity_type: str,
        deadline: str,
        summary: str,
        note: Optional[str] = None,
        assigned_user_id: Optional[int] = None,
    ) -> dict:
        try:
            types = client.execute(
                "mail.activity.type", "search_read",
                [["name", "ilike", activity_type]],
                fields=["id", "name"], limit=1,
            )
            if not types:
                return {"error": f"Activity type '{activity_type}' not found."}
            type_id = types[0]["id"]
            model_id = client.get_model_id(model)
            if not model_id:
                return {"error": f"Model '{model}' not found in ir.model."}
            vals: dict[str, Any] = {
                "activity_type_id": type_id,
                "date_deadline": deadline,
                "summary": summary,
                "res_id": record_id,
                "res_model_id": model_id,
            }
            if note:
                vals["note"] = note
            if assigned_user_id:
                vals["user_id"] = assigned_user_id
            activity_id = client.execute("mail.activity", "create", vals)
            return {"activity_id": activity_id, "activity_type": types[0]["name"], "deadline": deadline}
        except Exception as exc:
            return {"error": str(exc)}


# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------

def _is_odoo_command(value: Any) -> bool:
    """Return True if value is an Odoo Command list ([[int, ...], ...]).

    Odoo Commands are lists whose first element is itself a list/tuple whose
    first element is an integer command code (0–6).  Recognising them lets
    the update tool bypass the static readonly check and pass the value
    straight to write() — the server enforces the real access rules.
    """
    return (
        isinstance(value, list)
        and bool(value)
        and isinstance(value[0], (list, tuple))
        and bool(value[0])
        and isinstance(value[0][0], int)
    )


def _active_ctx(record_id: int, model: str) -> dict:
    """Standard active-record context injected when calling actions on a record."""
    return {
        "active_id": record_id,
        "active_ids": [record_id],
        "active_model": model,
    }


def _normalise(result: Any) -> dict:
    """Normalise Odoo method return values to a consistent dict.

    True / False / None all mean "action completed, stay on current form"
    in Odoo's return convention — no redirect action dict is produced.
    """
    if result is True or result is False or result is None:
        return {"success": True}
    if isinstance(result, dict):
        return result
    if isinstance(result, (int, str)):
        return {"result": result}
    return {"result": str(result)}


def _action_check(client: OdooClient, model: str, record_id: int) -> dict:
    """Re-read the record after a direct action (True/False/None return or None-marshal).

    Returns {success, record_after} so the caller has actual evidence of what
    changed in a single round-trip — the "action, check" pattern.
    Probes a small fixed set of common status fields filtered to those that
    actually exist on the model, so it is safe for both regular and transient
    wizard records.
    """
    _PROBE = ["state", "display_name", "name", "active"]
    try:
        valid = client.valid_field_names(model)
        fields = [f for f in _PROBE if f in valid] or ["display_name"]
        rows = client.execute(model, "read", [record_id], fields=fields)
        if rows:
            return {"success": True, "record_after": rows[0]}
    except Exception:
        pass
    return {"success": True}

File: src/tools/intent.ts
/** Layer 4 — Primary Intent tools: create, update, execute_action, archive,
 * post_message, schedule_activity. */
import { McpServer } from '@modelcontextprotocol/sdk/server/mcp.js';
import { z } from 'zod';
import { OdooClient } from '../odooClient.js';
import { Cache } from '../cache.js';
import { resolveContext, viewFieldNames, safeEvalDict } from './supporting.js';
import { collectModelActions } from './discovery.js';
import { ok } from '../utils.js';

function isOdooCommand(value: unknown): boolean {
  return Array.isArray(value) && value.length > 0
    && Array.isArray(value[0]) && value[0].length > 0
    && typeof value[0][0] === 'number';
}

function activeCtx(recordId: number, model: string): Record<string, unknown> {
  return { active_id: recordId, active_ids: [recordId], active_model: model };
}

function normalise(result: unknown): Record<string, unknown> {
  if (result === true || result === false || result === null || result === undefined) return { success: true };
  if (result && typeof result === 'object' && !Array.isArray(result)) return result as Record<string, unknown>;
  if (typeof result === 'number' || typeof result === 'string') return { result };
  return { result: String(result) };
}

async function actionCheck(client: OdooClient, model: string, recordId: number): Promise<Record<string, unknown>> {
  const PROBE = ['state', 'display_name', 'name', 'active'];
  try {
    const valid = await client.validFieldNames(model);
    const fields = PROBE.filter(f => valid.has(f));
    const rows = await client.execute(model, 'read', [[recordId]], { fields: fields.length ? fields : ['display_name'] }) as unknown[];
    if (rows.length) return { success: true, record_after: rows[0] };
  } catch { /* fall through */ }
  return { success: true };
}

export function register(server: McpServer, client: OdooClient, cache: Cache): void {

  server.registerTool(
    'create',
    {
      description:
        'Create a new record. values: dict of field/value pairs (form-view fields only). ' +
        'Defaults are merged with provided values automatically. ' +
        'Pass action_id to include the action\'s context (e.g. default_partner_id). ' +
        'Pass context dict directly for wizard models. ' +
        'Returns {id, display_name} or {error}.',
      inputSchema: {
        model: z.string(),
        values: z.record(z.unknown()),
        action_id: z.number().int().optional(),
        context: z.record(z.unknown()).optional(),
      },
    },
    async ({ model, values, action_id, context }) => {
      try {
        const mergedCtx = await resolveContext(client, cache, action_id, context);
        const newId = await client.execute(model, 'create', [values], { context: mergedCtx }) as number;
        const rows = await client.execute(model, 'read', [[newId]], { fields: ['id', 'display_name'] }) as unknown[];
        return ok(rows[0] ?? { id: newId });
      } catch (e) { return ok({ error: String(e) }); }
    },
  );

  server.registerTool(
    'update',
    {
      description:
        'Update fields on an existing record. values: {field: value, ...}. ' +
        'Writes both form-view fields and model fields not exposed in the form view. ' +
        'One2many / many2many fields accept Odoo Command tuples directly: ' +
        '[[0,0,{vals}]] create+link, [[1,id,{vals}]] update line, [[2,id]] delete line, [[6,0,[ids]]] replace set. ' +
        'Returns {success, updated_fields, non_form_fields} or {error}.',
      inputSchema: {
        model: z.string(),
        record_id: z.number().int(),
        values: z.record(z.unknown()),
      },
    },
    async ({ model, record_id, values }) => {
      try {
        const formFields = new Set(await viewFieldNames(client, cache, model, 'form'));
        const meta = await client.execute(model, 'fields_get', [], { attributes: ['readonly'] }) as Record<string, { readonly?: boolean }>;
        const writable: Record<string, unknown> = {};
        for (const [k, v] of Object.entries(values)) {
          if (k in meta && (isOdooCommand(v) || !meta[k]?.readonly)) {
            writable[k] = v;
          }
        }
        if (!Object.keys(writable).length) return ok({ error: 'No writable fields found in the provided values.' });
        await client.execute(model, 'write', [[record_id], writable]);
        const nonFormWritten = Object.keys(writable).filter(k => !formFields.has(k));
        const result: Record<string, unknown> = { success: true, updated_fields: Object.keys(writable) };
        if (nonFormWritten.length) result['non_form_fields'] = nonFormWritten;
        return ok(result);
      } catch (e) { return ok({ error: String(e) }); }
    },
  );

  server.registerTool(
    'execute_action',
    {
      description:
        'Execute a button or server action on a record. ' +
        'action: button name (method) or label as shown in get_model_actions — ' +
        'e.g. "action_confirm", "Confirm", "Privacy Lookup". ' +
        'View buttons (type=object) call the method directly; server actions use ir.actions.server.run. ' +
        'Returns the Odoo action result, {success: true}, or {error}.',
      inputSchema: {
        model: z.string(),
        record_id: z.number().int(),
        action: z.string(),
      },
    },
    async ({ model, record_id, action }) => {
      try {
        const actionMap = await collectModelActions(client, model);

        // 1. View buttons
        for (const btn of (actionMap['view_buttons'] as Array<Record<string, unknown>> ?? [])) {
          if (btn['name'] !== action && btn['label'] !== action) continue;

          if (btn['type'] === 'action') {
            const actionId = parseInt(String(btn['name']), 10);
            const meta = await client.execute('ir.actions.actions', 'read', [[actionId]], {
              fields: ['type', 'name'],
            }) as Array<{ type: string; name: string }>;
            if (!meta.length) return ok({ error: `Action id ${actionId} not found.` });
            const actionType = meta[0].type ?? 'ir.actions.act_window';
            const full = await client.execute(actionType, 'read', [[actionId]]) as Array<Record<string, unknown>>;
            if (!full.length) return ok({ error: `Could not load action ${actionId}.` });
            const actionDef = { ...full[0] };
            const ctx = safeEvalDict(actionDef['context']);
            Object.assign(ctx, activeCtx(record_id, model));
            actionDef['context'] = ctx;
            return ok(actionDef);
          }

          // type=object → call_button
          const result = await client.httpCall('/web/dataset/call_button', {
            model, method: btn['name'], args: [[record_id]], kwargs: {},
          });
          if (result === false || result === null || result === undefined) {
            return ok(await actionCheck(client, model, record_id));
          }
          return ok(normalise(result));
        }

        // 2. Server actions
        for (const sa of (actionMap['server_actions'] as Array<Record<string, unknown>> ?? [])) {
          if (String(sa['id']) !== action && sa['name'] !== action) continue;
          const result = await client.execute('ir.actions.server', 'run', [[sa['id']]], {
            context: activeCtx(record_id, model),
          });
          if (result === true || result === false || result === null || result === undefined) {
            return ok(await actionCheck(client, model, record_id));
          }
          return ok(normalise(result));
        }

        return ok({
          error: `Action '${action}' not found on '${model}'. Use get_model_actions() to list available actions.`,
        });
      } catch (e) { return ok({ error: String(e) }); }
    },
  );

  server.registerTool(
    'archive',
    {
      description:
        'Archive (deactivate) a record by setting active=False. ' +
        'Only works on models that have an active field (most standard models do). ' +
        'Returns {success: true} or {error}.',
      inputSchema: { model: z.string(), record_id: z.number().int() },
    },
    async ({ model, record_id }) => {
      try {
        const meta = await client.execute(model, 'fields_get', [], { attributes: ['type'] }) as Record<string, unknown>;
        if (!('active' in meta)) return ok({ error: `Model '${model}' has no active field and cannot be archived.` });
        await client.execute(model, 'write', [[record_id], { active: false }]);
        return ok({ success: true, record_id, active: false });
      } catch (e) { return ok({ error: String(e) }); }
    },
  );

  server.registerTool(
    'post_message',
    {
      description:
        'Post a message or internal note on a record (requires mail.thread). ' +
        'message_type: "comment" (sent to followers) or "note" (internal log note, not emailed). ' +
        'Returns {message_id} or {error}.',
      inputSchema: {
        model: z.string(),
        record_id: z.number().int(),
        body: z.string(),
        message_type: z.enum(['comment', 'note']).default('comment'),
      },
    },
    async ({ model, record_id, body, message_type }) => {
      try {
        const msgId = await client.execute(model, 'message_post', [[record_id]], {
          body,
          message_type: 'comment',
          subtype_xmlid: message_type === 'note' ? 'mail.mt_note' : 'mail.mt_comment',
        });
        return ok({ message_id: msgId });
      } catch (e) { return ok({ error: String(e) }); }
    },
  );

  server.registerTool(
    'schedule_activity',
    {
      description:
        'Schedule an activity on a record. ' +
        'activity_type: name of the activity type (e.g. "To-Do", "Email", "Phone Call"). ' +
        'deadline: ISO date string YYYY-MM-DD. ' +
        'summary: short title. note: longer description (optional). ' +
        'assigned_user_id: who to assign (default: current user). ' +
        'Returns {activity_id, activity_type, deadline} or {error}.',
      inputSchema: {
        model: z.string(),
        record_id: z.number().int(),
        activity_type: z.string(),
        deadline: z.string(),
        summary: z.string(),
        note: z.string().optional(),
        assigned_user_id: z.number().int().optional(),
      },
    },
    async ({ model, record_id, activity_type, deadline, summary, note, assigned_user_id }) => {
      try {
        const types = await client.execute('mail.activity.type', 'search_read',
          [[['name', 'ilike', activity_type]]],
          { fields: ['id', 'name'], limit: 1 },
        ) as Array<{ id: number; name: string }>;
        if (!types.length) return ok({ error: `Activity type '${activity_type}' not found.` });

        const modelId = await client.getModelId(model);
        if (!modelId) return ok({ error: `Model '${model}' not found in ir.model.` });

        const vals: Record<string, unknown> = {
          activity_type_id: types[0].id,
          date_deadline: deadline,
          summary,
          res_id: record_id,
          res_model_id: modelId,
        };
        if (note) vals['note'] = note;
        if (assigned_user_id) vals['user_id'] = assigned_user_id;

        const activityId = await client.execute('mail.activity', 'create', [vals]);
        return ok({ activity_id: activityId, activity_type: types[0].name, deadline });
      } catch (e) { return ok({ error: String(e) }); }
    },
  );
}
